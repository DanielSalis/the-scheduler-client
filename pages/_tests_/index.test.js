import { mount, createLocalVue } from '@vue/test-utils'
import Vuetify from 'vuetify'
import Vuex from 'vuex'
import IndexPage from '@/pages/index.vue' // Adjust the path as needed

// Create a local Vue instance
const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(Vuetify)

// Mock layout component
const DefaultLayoutMock = {
  template: '<div><slot /></div>',
}

// Mock any components used in the layout
const GNavigatorMock = {
  template: '<div></div>',
}

// Setup Vuex store with any necessary modules and state
const store = new Vuex.Store({
  modules: {
    auth: {
      namespaced: true,
      state: {
        getAuthData: { user_role: { name: 'Admin' } },
      },
      getters: {
        getAuthData: (state) => state.getAuthData,
      },
      mutations: {
        setAuthData: jest.fn(),
      },
    },
  },
})

// Define the Vuetify instance
let vuetify

describe('Index Page', () => {
  let wrapper

  beforeEach(() => {
    vuetify = new Vuetify()

    wrapper = mount(IndexPage, {
      localVue,
      vuetify,
      store,
      stubs: {
        DefaultLayout: DefaultLayoutMock,
        GNavigator: GNavigatorMock,
      },
      mocks: {
        $router: { push: jest.fn() },
        $cookies: {
          get: jest.fn(() => ({ user_role: { name: 'Admin' } })),
        },
      },
    })
  })

  it('renders the Index Page correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('renders the correct number of v-expansion-panels based on user role', () => {
    // Find all v-expansion-panel elements by class name
    const panels = wrapper.findAll('.v-expansion-panel');

    // Assert that the number of panels matches the length of availableRoutes
    expect(panels.length).toBe(wrapper.vm.availableRoutes.length);
  });

  describe("when user is a nurse", () => {
    it('renders no v-expansion-panels when user role is not Admin', async () => {
      // Update the store to return a non-Admin role
      store.state.auth.getAuthData = { user_role: { name: 'Enfermeiro' } }

      // Re-mount the component to reflect the updated store
      wrapper = mount(IndexPage, {
        localVue,
        vuetify,
        store,
        stubs: {
          DefaultLayout: DefaultLayoutMock,
          GNavigator: GNavigatorMock,
        },
        mocks: {
          $router: { push: jest.fn() },
          $cookies: {
            get: jest.fn(() => ({ user_role: { name: 'Enfermeiro' } })),
          },
        },
      })

      // Find all v-expansion-panel elements by class name
      const panels = wrapper.findAll('.v-expansion-panel');

      // Assert that no panels are rendered when the user role is not Admin
      expect(panels.length).toBe(2);
    });
  })

  describe("when user is a worker", () => {
    it('renders no v-expansion-panels when user role is not Admin', async () => {
      // Update the store to return a non-Admin role
      store.state.auth.getAuthData = { user_role: { name: 'Funcionario' } }

      // Re-mount the component to reflect the updated store
      wrapper = mount(IndexPage, {
        localVue,
        vuetify,
        store,
        stubs: {
          DefaultLayout: DefaultLayoutMock,
          GNavigator: GNavigatorMock,
        },
        mocks: {
          $router: { push: jest.fn() },
          $cookies: {
            get: jest.fn(() => ({ user_role: { name: 'Funcionario' } })),
          },
        },
      })

      // Find all v-expansion-panel elements by class name
      const panels = wrapper.findAll('.v-expansion-panel');

      // Assert that no panels are rendered when the user role is not Admin
      expect(panels.length).toBe(0);
    });
  })
})
