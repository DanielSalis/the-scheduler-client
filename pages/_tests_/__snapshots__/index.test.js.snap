// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index Page renders the Index Page correctly 1`] = `
<div
  class="g-container home-page__container"
>
  <div
    class="v-item-group theme--light v-expansion-panels"
  >
    <div
      aria-expanded="false"
      class="v-expansion-panel"
      style="width: 100%;"
    >
      <button
        class="v-expansion-panel-header text-h5 ml-3"
        type="button"
      >
        
        Admin
      
        <transition-stub
          name="fade-transition"
        >
          <div
            class="v-expansion-panel-header__icon"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate mdi mdi-chevron-down theme--light"
            />
          </div>
        </transition-stub>
      </button>
       
      <transition-stub
        name="expand-transition"
      >
        <!---->
      </transition-stub>
    </div>
    <div
      aria-expanded="false"
      class="v-expansion-panel"
      style="width: 100%;"
    >
      <button
        class="v-expansion-panel-header text-h5 ml-3"
        type="button"
      >
        
        Ferramentas
      
        <transition-stub
          name="fade-transition"
        >
          <div
            class="v-expansion-panel-header__icon"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate mdi mdi-chevron-down theme--light"
            />
          </div>
        </transition-stub>
      </button>
       
      <transition-stub
        name="expand-transition"
      >
        <!---->
      </transition-stub>
    </div>
    <div
      aria-expanded="false"
      class="v-expansion-panel"
      style="width: 100%;"
    >
      <button
        class="v-expansion-panel-header text-h5 ml-3"
        type="button"
      >
        
        Dashboards
      
        <transition-stub
          name="fade-transition"
        >
          <div
            class="v-expansion-panel-header__icon"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate mdi mdi-chevron-down theme--light"
            />
          </div>
        </transition-stub>
      </button>
       
      <transition-stub
        name="expand-transition"
      >
        <!---->
      </transition-stub>
    </div>
  </div>
</div>
`;
